src/main/java/org/folio/circulation/domain/validation/ItemLimitValidator.java

```diff
@@ -0,0 +1,119 @@
+    LoanAndRelatedRecords records) {
+
+    Loan loan = records.getLoan();
+    Integer itemLimit = loan.getLoanPolicy().getItemLimit();
+
+    if (itemLimit == null) {
+      return completedFuture(succeeded(records));
+    }
+
+    return loanPolicyRepository.lookupConditions(loan.getItem(), loan.getUser())
+      .thenComposeAsync(result -> result.failAfter(rule -> isLimitReached(rule, records),
+        conditions -> {
+          String message = getErrorMessage(conditions);
+          return itemLimitErrorFunction.apply(String.format("Patron has reached maximum item limit of %d items %s",
+            itemLimit, message));
+        }))
+      .thenApply(result -> result.map(v -> records));
+  }
+
+  private CompletableFuture<Result<Boolean>> isLimitReached(List<String> ruleConditions, LoanAndRelatedRecords records) {
+
+    if (!isRuleMaterialTypePresent(ruleConditions) && !isRuleLoanTypePresent(ruleConditions)) {
+      return ofAsync(() -> false);
+    }
+
+    Item item = records.getLoan().getItem();
+    String materialType = item.getMaterialType().getString("name");
+    String loanType = item.getLoanTypeName();
+    Integer itemLimit = records.getLoan().getLoanPolicy().getItemLimit();
+
+    if (isRuleMaterialTypePresent(ruleConditions) || isRuleLoanTypePresent(ruleConditions)) {
+      return loanRepository.findOpenLoansByUserIdAndLoanPolicyIdWithItem(records)
+        .thenApply(r -> r.map(loans -> loans.getRecords().stream()
+          .filter(loan -> isMaterialTypeMatchInRetrievedLoan(materialType, loan))
+          .filter(loan -> isLoanTypeMatchInRetrievedLoan(loanType, loan))
+          .count()))
+        .thenApply(r -> r.map(loansCount ->loansCount >= itemLimit));
+    }
+
+    return loanRepository.findOpenLoansByUserIdAndLoanPolicyId(records)
+      .thenApply(r -> r.map(loans -> loans.getTotalRecords() >= itemLimit));
+  }
+
+  private boolean isMaterialTypeMatchInRetrievedLoan(String expectedMaterialType, Loan loan) {
+    return expectedMaterialType.equalsIgnoreCase(loan.getItem().getMaterialType().getString("name"));
+  }
+
+  private boolean isLoanTypeMatchInRetrievedLoan(String expectedLoanType, Loan loan) {
+    return expectedLoanType.equalsIgnoreCase(loan.getItem().getLoanTypeName());
+  }
+
+  private boolean isRuleMaterialTypePresent(List<String> conditions) {
+    return conditions.contains("ItemType");
+  }
+
+  private boolean isRuleLoanTypePresent(List<String> conditions) {
+    return conditions.contains("LoanType");
+  }
+
+  private boolean isRulePatronGroupPresent(List<String> conditions) {
+    return conditions.contains("PatronGroup");
+  }
+
+  private String getErrorMessage(List<String> conditions) {
+    boolean isRuleMaterialTypePresent = isRuleMaterialTypePresent(conditions);
+    boolean isRuleLoanTypePresent = isRuleLoanTypePresent(conditions);
+    boolean isRulePatronGroupPresent = isRulePatronGroupPresent(conditions);
```

Comment:
Why is the patron group relevant? My understanding is that the patron group from the circulation rule does not have any effect on which items are considered within the limit?

Reply:
I need only check if a patron group present in circulation rule according to the requirements in [CIRC-558](https://issues.folio.org/browse/CIRC-558). If it is, I will form a specific error message according to the requirements.

Reply:
Ok, I've asked a follow up question about that on the issue.